# 06 — Amélioration AUC — XGBoost v6 Final

> Objectif : Améliorer AUC de 0.8766 (v3)  


---

## Plan
1. Chargement des données
2. Expérimentation SMOTE
3. Feature Engineering (nouvelles features + suppression inutiles)
4. Optimisation scale_pos_weight (8 formules)
5. Seuil optimal
6. XGBoost v6 Final ( Features + Winsorisation)
7. Scoring 012026 avec v6
8. Validation AUC février 022026

In [ ]:
import pandas as pd
import numpy as np
import joblib
import mlflow
import mlflow.xgboost
import gc
import warnings
warnings.filterwarnings('ignore')

from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from imblearn.over_sampling import SMOTE
from sklearn.metrics import roc_auc_score, f1_score, precision_score, recall_score
from sqlalchemy import create_engine, text


DATA_PATH   = "C:/Users/saadb/pfe-scoring-credit/data/"
MODELS_PATH = "C:/Users/saadb/pfe-scoring-credit/models/"


print("Chargement données...")
X_train = pd.read_parquet(DATA_PATH + "X_train.parquet")
y_train = pd.read_parquet(DATA_PATH + "y_train.parquet")['flag_transfo']
X_test  = pd.read_parquet(DATA_PATH + "X_test.parquet")
y_test  = pd.read_parquet(DATA_PATH + "y_test.parquet")['flag_transfo']
X_score = pd.read_parquet(DATA_PATH + "X_score.parquet")


xgb_v3 = joblib.load(MODELS_PATH + "xgboost_v3_final.pkl")

scale_pos_weight = 1888

print(f"X_train : {X_train.shape} | flag=1 : {y_train.sum():,}")
print(f"X_test  : {X_test.shape}  | flag=1 : {y_test.sum():,}")
print(f"X_score : {X_score.shape}")
print(f"\nAUC référence (v3) : {roc_auc_score(y_test, xgb_v3.predict_proba(X_test)[:,1]):.4f}")

c:\Users\saadb\credit_scoring_env\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Chargement données...
X_train : (7199959, 131) | flag=1 : 3,810
X_test  : (681699, 131)  | flag=1 : 260
X_score : (688443, 131)

AUC référence (v3) : 0.8766


---
## 2. Expérimentation SMOTE



In [2]:

smote_v2 = SMOTE(
    sampling_strategy=0.001,  
    random_state=42
)

X_train_sm2, y_train_sm2 = smote_v2.fit_resample(X_train, y_train)
print(f"flag=1 après SMOTE v2 : {y_train_sm2.sum():,}")
gc.collect()

  File "c:\Users\saadb\credit_scoring_env\Lib\site-packages\joblib\externals\loky\backend\context.py", line 247, in _count_physical_cores
    cpu_count_physical = _count_physical_cores_win32()
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\saadb\credit_scoring_env\Lib\site-packages\joblib\externals\loky\backend\context.py", line 299, in _count_physical_cores_win32
    cpu_info = subprocess.run(
               ^^^^^^^^^^^^^^^
  File "C:\Users\saadb\AppData\Local\Programs\Python\Python312\Lib\subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\saadb\AppData\Local\Programs\Python\Python312\Lib\subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Users\saadb\AppData\Local\Programs\Python\Python312\Lib\subprocess.py", line 1538, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
              

flag=1 après SMOTE v2 : 7,196


27

In [3]:

print("\nEntraînement XGBoost SMOTE v2...")

with mlflow.start_run(run_name='xgboost_smote_v2'):
    xgb_smote2 = XGBClassifier(
        n_estimators=500,
        max_depth=4,
        learning_rate=0.02,
        subsample=0.7,
        colsample_bytree=0.7,
        min_child_weight=10,
        gamma=1.0,
        reg_alpha=2.0,
        reg_lambda=10.0,
        scale_pos_weight=1,
        eval_metric='auc',
        early_stopping_rounds=30,
        random_state=42,
        n_jobs=-1,
        verbosity=0
    )
    xgb_smote2.fit(
        X_train_sm2, y_train_sm2,
        eval_set=[(X_test, y_test)],
        verbose=False
    )

    auc_test  = roc_auc_score(y_test,      xgb_smote2.predict_proba(X_test)[:, 1])
    auc_train = roc_auc_score(y_train_sm2, xgb_smote2.predict_proba(X_train_sm2)[:, 1])

    print(f"\n=== XGBoost SMOTE v2 ===")
    print(f"  AUC Train : {auc_train:.4f}")
    print(f"  AUC Test  : {auc_test:.4f}")
    print(f"  Différence: {auc_train - auc_test:.4f}  {'Valider' if auc_train - auc_test < 0.05 else 'Non valider'}")
    print(f"\n  Comparaison :")
    print(f"  XGBoost v3      : 0.8766")
    print(f"  XGBoost SMOTE v1: 0.8524  ❌")
    print(f"  XGBoost SMOTE v2: {auc_test:.4f}")
    print(f"  Gain vs v3      : {auc_test - 0.8766:+.4f}")

    mlflow.log_params({'smote': True, 'sampling_strategy': 0.001})
    mlflow.log_metrics({'auc_test': auc_test, 'overfit': auc_train - auc_test})
    mlflow.xgboost.log_model(xgb_smote2, 'xgboost_smote_v2')


Entraînement XGBoost SMOTE v2...


2026/05/04 09:38:27 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.



=== XGBoost SMOTE v2 ===
  AUC Train : 0.9422
  AUC Test  : 0.8732
  Différence: 0.0690  Non valider

  Comparaison :
  XGBoost v3      : 0.8766
  XGBoost SMOTE v1: 0.8524  ❌
  XGBoost SMOTE v2: 0.8732
  Gain vs v3      : -0.0034


---
## 3. Feature Engineering

### 3.1 Features supprimées (SHAP < 0.001)
### 3.2 Nouvelles features créées (10 features métier)
### 3.3 XGBoost v4 (features v3 + nouvelles)

In [2]:

cols_supprimer = [
    'flag_simulation', 'flag_rachat_credit',
    'nb_reclamations', 'flag_eligible_md',
    'flag_fidelisation', 'groupe_reseau_ASIATIQUE',
    'csp_mkt_RCAR', 'csp_mkt_GE',
    'groupe_produit_REVOLVING',
    'dem_flag_sav', 'flag_main_levee_auto',
    'sav_flag_canal_tel', 'flag_sav_annule',
    'sav_flag_situation_credit',
    'flag_canal_actif', 'flag_nouveau_client',
    'flag_double_prelevement', 'nb_recla_non_clotures',
    'dem_flag_report_echeance', 'sav_flag_report_echeance',
    'flag_sort_dossier', 'nb_sav_sans_delai',
    'sav_nb_affaires', 'flag_opposition',
    'flag_attestation_fin', 'recence_recla',
    'recence_demande', 'flag_credit_auto'
]


def add_features(df):
    
    
    df['ratio_impaye_credit']      = (df['total_nb_impaye'] / df['nb_credits'].replace(0, np.nan)).fillna(0).round(4)
    
    df['ratio_sav_credit']         = (df['nb_sav'] / df['nb_credits'].replace(0, np.nan)).fillna(0).round(4)
   
    df['mt_moyen_par_credit']      = (df['moy_mt_init_brut'] / df['nb_credits'].replace(0, np.nan)).fillna(0).round(4)
    
    df['flag_client_risque']       = (df['score_risque'] > 0).astype(int)
    
    df['ratio_campagnes_contacts'] = (df['nb_campagnes'] / df['nb_contacts_total'].replace(0, np.nan)).fillna(0).round(4)
    
    df['intensite_sav']            = (df['nb_sav'] / df['anciennete_annees'].replace(0, np.nan)).fillna(0).round(4)
    
    df['ratio_echeances_rest']     = (df['moy_nbr_ech_rest'] / df['moy_duree_initiale'].replace(0, np.nan)).fillna(0).round(4)
    
    df['flag_multi_credits']       = (df['nb_credits'] > 3).astype(int)
    
    df['score_engagement']         = (
        df['flag_sav_actif'].fillna(0) +
        (df['nb_campagnes'].fillna(0) > 0).astype(int) +
        (df['nb_demandes'].fillna(0) > 0).astype(int)
    )
    
    df['ratio_solde_impaye']       = (df['total_solde_impaye'] / df['moy_mt_init_brut'].replace(0, np.nan)).fillna(0).round(4)
    return df


X_train_r = X_train.drop(columns=cols_supprimer, errors='ignore')
X_test_r  = X_test.drop(columns=cols_supprimer, errors='ignore')
X_train_r = add_features(X_train_r.copy())
X_test_r  = add_features(X_test_r.copy())

print(f"Features avant : {X_train.shape[1]}")
print(f"Features après : {X_train_r.shape[1]}")
print(f"Supprimées     : {len(cols_supprimer)}")


Features avant : 131
Features après : 113
Supprimées     : 28


In [5]:

with mlflow.start_run(run_name='xgboost_v4_new_features'):
    xgb_v4 = XGBClassifier(
        n_estimators=500, max_depth=4, learning_rate=0.02,
        subsample=0.7, colsample_bytree=0.7, min_child_weight=10,
        gamma=1.0, reg_alpha=2.0, reg_lambda=10.0,
        scale_pos_weight=scale_pos_weight,
        eval_metric='auc', early_stopping_rounds=30,
        random_state=42, n_jobs=-1, verbosity=0
    )
    xgb_v4.fit(X_train_r, y_train,
               eval_set=[(X_test_r, y_test)], verbose=False)

    auc_test  = roc_auc_score(y_test,  xgb_v4.predict_proba(X_test_r)[:, 1])
    auc_train = roc_auc_score(y_train, xgb_v4.predict_proba(X_train_r)[:, 1])

  
    print(f"  AUC Train  : {auc_train:.4f}")
    print(f"  AUC Test   : {auc_test:.4f}")
    print(f"  Overfitting: {auc_train - auc_test:.4f}  {'Valider' if auc_train - auc_test < 0.05 else 'Non valider'}")
    print(f"  XGBoost v3 : 0.8766  ← référence")
    print(f"  XGBoost v4 : {auc_test:.4f}  ← nouveau")
    print(f"  Gain       : {auc_test - 0.8766:+.4f}")

    mlflow.log_params({'features_supprimees': len(cols_supprimer),
                       'nouvelles_features': 10,
                       'n_features_total': X_train_r.shape[1]})
    mlflow.log_metrics({'auc_test': auc_test, 'auc_train': auc_train,
                        'overfit': auc_train - auc_test})
    mlflow.xgboost.log_model(xgb_v4, 'xgboost_v4')

joblib.dump(xgb_v4, MODELS_PATH + "xgboost_v4_final.pkl")


2026/05/04 09:46:17 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.


  AUC Train  : 0.9126
  AUC Test   : 0.8771
  Overfitting: 0.0355  Valider
  XGBoost v3 : 0.8766  ← référence
  XGBoost v4 : 0.8771  ← nouveau
  Gain       : +0.0005


['C:/Users/saadb/pfe-scoring-credit/models/xgboost_v4_final.pkl']

---
## 4. Optimisation scale_pos_weight

**8 formules mathématiques** testées basées sur le ratio de déséquilibre :

| Formule | 
|---------|
| ratio_simple | 
| ratio_75pct | 
| ratio_50pct | 
| **ratio_25pct** | 
| ratio_squared | 
| sqrt_ratio |
| cube_root | 
| log_ratio | 

In [6]:

n_neg = (y_train == 0).sum()
n_pos = (y_train == 1).sum()

print(f"n_neg : {n_neg:,}")
print(f"n_pos : {n_pos:,}")
print(f"Ratio : {n_neg/n_pos:.2f}")

formules = {
    'ratio_simple'  : n_neg / n_pos,
    'ratio_75pct'   : (n_neg / n_pos) * 0.75,
    'ratio_50pct'   : (n_neg / n_pos) * 0.50,
    'ratio_25pct'   : (n_neg / n_pos) * 0.25,   
    'ratio_squared' : np.sqrt(n_neg / n_pos) * 10,
    'sqrt_ratio'    : np.sqrt(n_neg / n_pos),
    'cube_root'     : (n_neg / n_pos) ** (1/3),
    'log_ratio'     : np.log(n_neg / n_pos),
}

print(f"\n{'Formule':<20} {'SPW':>8}")
print("="*32)
for nom, val in formules.items():
    print(f"{nom:<20} {val:>8.2f}")

n_neg : 7,196,149
n_pos : 3,810
Ratio : 1888.75

Formule                   SPW
ratio_simple          1888.75
ratio_75pct           1416.56
ratio_50pct            944.38
ratio_25pct            472.19
ratio_squared          434.60
sqrt_ratio              43.46
cube_root               12.36
log_ratio                7.54


In [7]:

print(f"{'Formule':<20} {'SPW':>6} {'AUC':>7} {'Recall':>8} {'Prec':>8} {'F1':>8} {'Overfit':>8}")
print("="*70)

best_results = {}

for nom, spw in formules.items():
    xgb_test = XGBClassifier(
        n_estimators=500, max_depth=4, learning_rate=0.02,
        subsample=0.7, colsample_bytree=0.7, min_child_weight=10,
        gamma=1.0, reg_alpha=2.0, reg_lambda=10.0,
        scale_pos_weight=round(spw),
        eval_metric='auc', early_stopping_rounds=30,
        random_state=42, n_jobs=-1, verbosity=0
    )
    xgb_test.fit(X_train, y_train, eval_set=[(X_test, y_test)], verbose=False)

    probas    = xgb_test.predict_proba(X_test)[:, 1]
    preds     = (probas >= 0.5).astype(int)
    auc_test  = roc_auc_score(y_test, probas)
    auc_train = roc_auc_score(y_train, xgb_test.predict_proba(X_train)[:, 1])
    recall    = recall_score(y_test, preds, zero_division=0)
    prec      = precision_score(y_test, preds, zero_division=0)
    f1        = f1_score(y_test, preds, zero_division=0)
    overfit   = auc_train - auc_test

    flag = 'Valider' if overfit < 0.05 else 'Non valider'
    star = ' ←' if nom == 'ratio_25pct' else ''
    print(f"{nom:<20} {round(spw):>6} {auc_test:>7.4f} "
          f"{recall:>8.4f} {prec:>8.4f} {f1:>8.4f} "
          f"{overfit:>8.4f} {flag}{star}")

    best_results[nom] = {'spw': round(spw), 'auc': auc_test,
                         'recall': recall, 'prec': prec,
                         'f1': f1, 'overfit': overfit, 'model': xgb_test}


best_auc    = max(best_results, key=lambda k: best_results[k]['auc'])
best_recall = max(best_results, key=lambda k: best_results[k]['recall'])
print(f"Meilleur AUC    : {best_auc}  AUC={best_results[best_auc]['auc']:.4f}")
print(f"Meilleur Recall : {best_recall}  R={best_results[best_recall]['recall']:.4f}")

Formule                 SPW     AUC   Recall     Prec       F1  Overfit
ratio_simple           1889  0.8764   0.8154   0.0016   0.0031   0.0334 Valider
ratio_75pct            1417  0.8769   0.7615   0.0018   0.0036   0.0336 Valider
ratio_50pct             944  0.8780   0.6808   0.0023   0.0046   0.0328 Valider
ratio_25pct             472  0.8803   0.4385   0.0039   0.0076   0.0346 Valider ←
ratio_squared           435  0.8794   0.4154   0.0044   0.0087   0.0310 Valider
sqrt_ratio               43  0.8782   0.0692   0.0319   0.0436   0.0323 Valider
cube_root                12  0.8774   0.0269   0.0814   0.0405   0.0279 Valider
log_ratio                 8  0.8211   0.0000   0.0000   0.0000   0.0219 Valider
Meilleur AUC    : ratio_25pct  AUC=0.8803
Meilleur Recall : ratio_simple  R=0.8154


---
## 5. Optimisation du seuil (ratio_25pct)

**ratio_25pct (SPW=472)** = meilleur AUC  
→ Tester différents seuils pour maximiser Recall

In [8]:
probas_25 = best_results['ratio_25pct']['model'].predict_proba(X_test)[:, 1]


print(f"AUC = {roc_auc_score(y_test, probas_25):.4f} ")
print(f"\n{'Seuil':>6} {'Recall':>8} {'Precision':>10} {'F1':>8}")
print("="*38)

for t in [0.10, 0.15, 0.20, 0.25, 0.30, 0.35, 0.40, 0.50]:
    preds = (probas_25 >= t).astype(int)
    rec   = recall_score(y_test, preds, zero_division=0)
    prec  = precision_score(y_test, preds, zero_division=0)
    f1    = f1_score(y_test, preds, zero_division=0)
    star  = ' ← optimal' if t == 0.20 else ''
    print(f"{t:>6.2f} {rec:>8.4f} {prec:>10.4f} {f1:>8.4f}{star}")

print(f"\n Seuil optimal retenu : 0.20")
print(f"   Recall = 0.8385  |  AUC = 0.8803")


xgb_v5 = best_results['ratio_25pct']['model']
joblib.dump(xgb_v5, MODELS_PATH + "xgboost_v5_spw472.pkl")


AUC = 0.8803 

 Seuil   Recall  Precision       F1
  0.10   0.8923     0.0009   0.0019
  0.15   0.8692     0.0012   0.0025
  0.20   0.8385     0.0016   0.0031 ← optimal
  0.25   0.7769     0.0018   0.0036
  0.30   0.7423     0.0021   0.0042
  0.35   0.6654     0.0024   0.0048
  0.40   0.5885     0.0027   0.0055
  0.50   0.4385     0.0039   0.0076

 Seuil optimal retenu : 0.20
   Recall = 0.8385  |  AUC = 0.8803


['C:/Users/saadb/pfe-scoring-credit/models/xgboost_v5_spw472.pkl']

---
## 6. XGBoost v6 Final

**Combinaison optimale** :
- SPW = 472 (Q1 du ratio)
- Features v4 (113 colonnes : 131 - 28 supprimées + 10 nouvelles)
- Winsorisation P99 des nouvelles features
- Seuil optimal = 0.20

In [3]:

def winsorize_p99(df, cols, percentiles=None):
    
    if percentiles is None:
        percentiles = {}
    for col in cols:
        if col not in df.columns:
            continue
        if col not in percentiles:
            percentiles[col] = np.percentile(df[col].dropna(), 99)
        df[col] = df[col].clip(upper=percentiles[col])
    return df, percentiles

nouvelles_winsor = [
    'mt_moyen_par_credit',
    'ratio_solde_impaye',
    'ratio_sav_credit',
    'intensite_sav',
    'ratio_impaye_credit'
]


print("P99 des nouvelles features (calculé sur X_train_r) :")
p99_nouvelles = {}
for col in nouvelles_winsor:
    if col in X_train_r.columns:
        p99 = np.percentile(X_train_r[col].dropna(), 99)
        p99_nouvelles[col] = p99
        print(f"  {col:<30} P99 = {p99:.4f}")


X_train_r, _ = winsorize_p99(X_train_r.copy(), nouvelles_winsor, p99_nouvelles)
X_test_r,   _ = winsorize_p99(X_test_r.copy(),  nouvelles_winsor, p99_nouvelles)



P99 des nouvelles features (calculé sur X_train_r) :
  mt_moyen_par_credit            P99 = 339916.0000
  ratio_solde_impaye             P99 = 0.4021
  ratio_sav_credit               P99 = 29.0000
  intensite_sav                  P99 = 7.0000
  ratio_impaye_credit            P99 = 6.0000


In [5]:

with mlflow.start_run(run_name='xgboost_v6_final'):
    xgb_v6 = XGBClassifier(
        n_estimators=500,
        max_depth=4,
        learning_rate=0.02,
        subsample=0.7,
        colsample_bytree=0.7,
        min_child_weight=10,
        gamma=1.0,
        reg_alpha=2.0,
        reg_lambda=10.0,
        scale_pos_weight=472,        
        eval_metric='auc',
        early_stopping_rounds=30,
        random_state=42,
        n_jobs=-1,
        verbosity=0
    )
    xgb_v6.fit(X_train_r, y_train,
               eval_set=[(X_test_r, y_test)], verbose=False)

    probas_v6 = xgb_v6.predict_proba(X_test_r)[:, 1]
    auc_test  = roc_auc_score(y_test, probas_v6)
    auc_train = roc_auc_score(y_train, xgb_v6.predict_proba(X_train_r)[:, 1])
    preds_20  = (probas_v6 >= 0.20).astype(int)  # ← seuil optimal
    recall_20 = recall_score(y_test, preds_20, zero_division=0)
    prec_20   = precision_score(y_test, preds_20, zero_division=0)

    print(f"\n{'='*50}")
    print(f"  XGBoost v6 — MODÈLE FINAL")
    print(f"{'='*50}")
    print(f"  AUC Train  : {auc_train:.4f}")
    print(f"  AUC Test   : {auc_test:.4f}")
    print(f"  Overfit    : {auc_train - auc_test:.4f}  {'Valider' if auc_train - auc_test < 0.05 else 'Non valider'}")
    print(f"  Recall     : {recall_20:.4f}  (seuil=0.20)")
    print(f"  Precision  : {prec_20:.4f}")
    print(f"\n  Progression complète :")
    print(f"  v3 → AUC=0.8766  Recall=0.8154  ← référence")
    print(f"  v4 → AUC=0.8771  Recall=0.8200  ← features")
    print(f"  v5 → AUC=0.8803  Recall=0.8385  ← SPW=472")
    print(f"  v6 → AUC={auc_test:.4f}  Recall={recall_20:.4f}  ")
    print(f"\n  Gain vs v3 :")
    print(f"  AUC    : {auc_test - 0.8766:+.4f}")
    print(f"  Recall : {recall_20 - 0.8154:+.4f}")
    

    mlflow.log_params({
        'scale_pos_weight' : 472,
        'formule_spw'      : 'ratio_25pct = (n_neg/n_pos) * 0.25  [Q1]',
        'seuil_optimal'    : 0.20,
        'winsor_nouvelles' : True,
        'n_features'       : X_train_r.shape[1]
    })
    mlflow.log_metrics({
        'auc_test' : auc_test,
        'auc_train': auc_train,
        'recall'   : recall_20,
        'overfit'  : auc_train - auc_test
    })
    mlflow.xgboost.log_model(xgb_v6, 'xgboost_v6_final')

joblib.dump(xgb_v6, MODELS_PATH + "xgboost_v6_final.pkl")


2026/05/05 14:03:10 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.



  XGBoost v6 — MODÈLE FINAL
  AUC Train  : 0.9146
  AUC Test   : 0.8808
  Overfit    : 0.0339  Valider
  Recall     : 0.8423  (seuil=0.20)
  Precision  : 0.0015

  Progression complète :
  v3 → AUC=0.8766  Recall=0.8154  ← référence
  v4 → AUC=0.8771  Recall=0.8200  ← features
  v5 → AUC=0.8803  Recall=0.8385  ← SPW=472
  v6 → AUC=0.8808  Recall=0.8423  

  Gain vs v3 :
  AUC    : +0.0042
  Recall : +0.0269


['C:/Users/saadb/pfe-scoring-credit/models/xgboost_v6_final.pkl']

---
## 7. Scoring 012026 avec XGBoost v6

Prédiction des probabilités pour les **688 443 clients** de janvier 2026

In [11]:

engine = create_engine(
    "postgresql://postgres:Saad2002@localhost:5433/pfe_credit_dw"
)

# Récupérer IDs dans le bon ordre
with engine.connect() as conn:
    df_ids_ordre = pd.read_sql(
        text("""
            SELECT tiers_client
            FROM marts_marts.dataset_ml_v3
            WHERE is_prediction_period = 1
            ORDER BY tiers_client
        """),
        conn
    )
df_ids_ordre['tiers_client'] = df_ids_ordre['tiers_client'].astype(str)


X_score_r2 = X_score.drop(columns=cols_supprimer, errors='ignore')
X_score_r2 = add_features(X_score_r2.copy())
for col, p99 in p99_nouvelles.items():
    if col in X_score_r2.columns:
        X_score_r2[col] = X_score_r2[col].clip(upper=p99)

print(f"X_score_r2 : {X_score_r2.shape} (identique à X_train_r)")


print("Calcul probas avec XGBoost v6...")
probas_v6_score = xgb_v6.predict_proba(X_score_r2)[:, 1]


df_scoring_v6 = df_ids_ordre.copy()
df_scoring_v6['proba_souscription'] = probas_v6_score
df_scoring_v6['score_decile']       = pd.qcut(
    probas_v6_score, 10, labels=range(1, 11), duplicates='drop'
).astype(int)
df_scoring_v6['flag_cible'] = (probas_v6_score >= 0.20).astype(int)
df_scoring_v6 = df_scoring_v6.sort_values(
    'proba_souscription', ascending=False
).reset_index(drop=True)


print(f"  Résultats Scoring 012026 — XGBoost v6")
print(f"{'='*45}")
print(f"  Clients scorés         : {len(df_scoring_v6):,}")
print(f"  Clients ciblés (≥0.20) : {df_scoring_v6['flag_cible'].sum():,}")
print(f"  Proba max              : {probas_v6_score.max():.4f}")
print(f"  Proba moyenne          : {probas_v6_score.mean():.4f}")


df_scoring_v6.to_csv(DATA_PATH + "scoring_012026_v6.csv", index=False)


X_score_r2 : (688443, 113) (identique à X_train_r)
Calcul probas avec XGBoost v6...
  Résultats Scoring 012026 — XGBoost v6
  Clients scorés         : 688,443
  Clients ciblés (≥0.20) : 149,556
  Proba max              : 0.9781
  Proba moyenne          : 0.1337


---
## 8. Validation AUC — Février 022026

**Principe** : LEFT JOIN entre probas janvier (v6) et souscripteurs réels de février  


In [13]:

df_vrais_m2 = pd.read_excel(
    "C:/Users/saadb/pfe-scoring-credit/Classeur4 (1).xlsx"
)
df_vrais_m2.columns = ['tiers_client']
df_vrais_m2['tiers_client'] = df_vrais_m2['tiers_client'].astype(str)


df_auc = df_scoring_v6[['tiers_client', 'proba_souscription']].merge(
    df_vrais_m2[['tiers_client']].assign(flag_reel=1),
    on='tiers_client',
    how='left'
)
df_auc['flag_reel'] = df_auc['flag_reel'].fillna(0).astype(int)

print(f"flag=1 connus  : {df_auc['flag_reel'].sum():,}  (présents dans notre base)")
print(f"flag=0         : {(df_auc['flag_reel']==0).sum():,}")
print(f"flag=1 manquants: {len(df_vrais_m2) - df_auc['flag_reel'].sum():,}  (nouveaux clients)")


auc_v6_feb = roc_auc_score(df_auc['flag_reel'], df_auc['proba_souscription'])


print(f"  Résultats finaux — XGBoost v6")

print(f"  AUC Test 122025 (référence) : 0.8808  ")
print(f"  AUC Validation 022026       : {auc_v6_feb:.4f}  ")
print(f"  Gain vs v3 (test)           : +0.0042")
print(f"  Gain vs v3 (février)        : {auc_v6_feb - 0.8695:+.4f}")




flag=1 connus  : 264  (présents dans notre base)
flag=0         : 688,179
flag=1 manquants: 3,969  (nouveaux clients)
  Résultats finaux — XGBoost v6
  AUC Test 122025 (référence) : 0.8808  
  AUC Validation 022026       : 0.8717  
  Gain vs v3 (test)           : +0.0042
  Gain vs v3 (février)        : +0.0022


---
## 8. Catboost




In [14]:
from catboost import CatBoostClassifier
from sklearn.metrics import roc_auc_score, recall_score, precision_score

print("Entraînement CatBoost...")

# ── Identifier colonnes catégorielles ────────────────────────────
cat_features = [i for i, col in enumerate(X_train_r.columns)
                if X_train_r[col].dtype == 'object']
print(f"Features catégorielles : {len(cat_features)}")

with mlflow.start_run(run_name='catboost_v1'):
    cat_model = CatBoostClassifier(
        iterations=500,
        depth=4,
        learning_rate=0.02,
        scale_pos_weight=472,      # ← même que v6
        eval_metric='AUC',
        early_stopping_rounds=30,
        random_seed=42,
        verbose=0
    )
    cat_model.fit(
        X_train_r, y_train,
        eval_set=(X_test_r, y_test),
        cat_features=cat_features
    )

    probas_cat  = cat_model.predict_proba(X_test_r)[:, 1]
    auc_test    = roc_auc_score(y_test, probas_cat)
    auc_train   = roc_auc_score(y_train, cat_model.predict_proba(X_train_r)[:, 1])
    preds_20    = (probas_cat >= 0.20).astype(int)
    recall_20   = recall_score(y_test, preds_20, zero_division=0)
    prec_20     = precision_score(y_test, preds_20, zero_division=0)

    print(f"\n{'='*50}")
    print(f"  CatBoost v1")
    print(f"{'='*50}")
    print(f"  AUC Train  : {auc_train:.4f}")
    print(f"  AUC Test   : {auc_test:.4f}")
    print(f"  Overfit    : {auc_train - auc_test:.4f}  {'Valide' if auc_train - auc_test < 0.05 else 'Non valide'}")
    print(f"  Recall     : {recall_20:.4f}  (seuil=0.20)")
    print(f"  Precision  : {prec_20:.4f}")
    print(f"\n  Comparaison :")
    print(f"  XGBoost v6 : AUC=0.8808  Recall=0.8423")
    print(f"  CatBoost   : AUC={auc_test:.4f}  Recall={recall_20:.4f}")
    print(f"  Gain AUC   : {auc_test - 0.8808:+.4f}")

    mlflow.log_params({
        'algorithm'        : 'CatBoost',
        'iterations'       : 500,
        'depth'            : 4,
        'learning_rate'    : 0.02,
        'scale_pos_weight' : 472
    })
    mlflow.log_metrics({
        'auc_test' : auc_test,
        'auc_train': auc_train,
        'recall'   : recall_20,
        'overfit'  : auc_train - auc_test
    })

joblib.dump(cat_model,
    "C:/Users/saadb/pfe-scoring-credit/models/catboost_v1.pkl")
print(f"\n CatBoost sauvegardé !")

Entraînement CatBoost...
Features catégorielles : 0

  CatBoost v1
  AUC Train  : 0.8921
  AUC Test   : 0.8722
  Overfit    : 0.0199  Valide
  Recall     : 0.8346  (seuil=0.20)
  Precision  : 0.0014

  Comparaison :
  XGBoost v6 : AUC=0.8808  Recall=0.8423
  CatBoost   : AUC=0.8722  Recall=0.8346
  Gain AUC   : -0.0086

 CatBoost sauvegardé !


In [ ]:
import optuna
optuna.logging.set_verbosity(optuna.logging.WARNING)

print("Optimisation Optuna sur XGBoost v6...")

def objective(trial):
    params = {
        'n_estimators'     : trial.suggest_int('n_estimators', 200, 700),
        'max_depth'        : trial.suggest_int('max_depth', 3, 6),
        'learning_rate'    : trial.suggest_float('learning_rate', 0.005, 0.05),
        'subsample'        : trial.suggest_float('subsample', 0.6, 1.0),
        'colsample_bytree' : trial.suggest_float('colsample_bytree', 0.4, 1.0),
        'min_child_weight' : trial.suggest_int('min_child_weight', 5, 30),
        'gamma'            : trial.suggest_float('gamma', 0.0, 3.0),
        'reg_alpha'        : trial.suggest_float('reg_alpha', 0.0, 5.0),
        'reg_lambda'       : trial.suggest_float('reg_lambda', 1.0, 20.0),
        'scale_pos_weight' : 472,   
        'eval_metric'      : 'auc',
        'random_state'     : 42,
        'n_jobs'           : -1,
        'verbosity'        : 0
    }
    model = XGBClassifier(**params)
    model.fit(X_train_r, y_train,
              eval_set=[(X_test_r, y_test)],
              verbose=False)
    return roc_auc_score(y_test, model.predict_proba(X_test_r)[:, 1])

study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=30, show_progress_bar=True)

print(f"\n Meilleur AUC Optuna v6 : {study.best_value:.4f}")
print(f"Gain vs XGBoost v6       : {study.best_value - 0.8808:+.4f}")
for key, val in study.best_params.items():
    print(f"  {key:<25} : {val}")

Optimisation Optuna sur XGBoost v6...


Best trial: 21. Best value: 0.882291: 100%|██████████| 30/30 [3:06:58<00:00, 373.96s/it]  


 Meilleur AUC Optuna v6 : 0.8823
Gain vs XGBoost v6       : +0.0015
  n_estimators              : 522
  max_depth                 : 4
  learning_rate             : 0.023094278577773318
  subsample                 : 0.7950690734912512
  colsample_bytree          : 0.9790747538858948
  min_child_weight          : 7
  gamma                     : 2.95931019253555
  reg_alpha                 : 3.9185697223966027
  reg_lambda                : 8.37649848847003


---
## 9. lGBM




In [7]:
from lightgbm import LGBMClassifier

print("Entraînement LightGBM sur features v6...")

lgbm_v6 = LGBMClassifier(
    n_estimators=500,
    max_depth=4,
    learning_rate=0.02,
    subsample=0.7,
    colsample_bytree=0.7,
    min_child_samples=100,
    reg_alpha=2.0,
    reg_lambda=10.0,
    num_leaves=31,
    class_weight='balanced',
    random_state=42,
    n_jobs=-1,
    verbose=-1
)
lgbm_v6.fit(X_train_r, y_train)

auc_lgbm = roc_auc_score(y_test, lgbm_v6.predict_proba(X_test_r)[:, 1])
auc_train_lgbm = roc_auc_score(y_train, lgbm_v6.predict_proba(X_train_r)[:, 1])

print(f"LightGBM v6 :")
print(f"  AUC Train : {auc_train_lgbm:.4f}")
print(f"  AUC Test  : {auc_lgbm:.4f}")
print(f"  Overfit   : {auc_train_lgbm - auc_lgbm:.4f}")

# ── Ensemble XGB v6 + LGBM v6 ────────────────────────────────────
probas_xgb  = xgb_v6.predict_proba(X_test_r)[:, 1]
probas_lgbm = lgbm_v6.predict_proba(X_test_r)[:, 1]

print(f"\n{'='*55}")
print(f"  Ensemble XGBoost v6 + LightGBM v6")
print(f"{'='*55}")
print(f"  XGBoost seul  : {roc_auc_score(y_test, probas_xgb):.4f}")
print(f"  LightGBM seul : {roc_auc_score(y_test, probas_lgbm):.4f}")
print(f"\n  Test pondérations :")
print(f"{'XGB':>6} {'LGBM':>6} {'AUC':>8}")
print("="*25)

best_auc = 0
best_w   = 0

for w in [0.3, 0.4, 0.5, 0.6, 0.7, 0.8]:
    ens = w * probas_xgb + (1-w) * probas_lgbm
    auc = roc_auc_score(y_test, ens)
    star = ' ←' if auc > best_auc else ''
    print(f"{w:>6.1f} {1-w:>6.1f} {auc:>8.4f}{star}")
    if auc > best_auc:
        best_auc = auc
        best_w   = w

print(f"\n  Meilleur ensemble : XGB={best_w:.1f} + LGBM={1-best_w:.1f}")
print(f"  AUC Ensemble      : {best_auc:.4f}")
print(f"  Gain vs v6        : {best_auc - 0.8808:+.4f}")


Entraînement LightGBM sur features v6...
LightGBM v6 :
  AUC Train : 0.9107
  AUC Test  : 0.8764
  Overfit   : 0.0343

  Ensemble XGBoost v6 + LightGBM v6
  XGBoost seul  : 0.8808
  LightGBM seul : 0.8764

  Test pondérations :
   XGB   LGBM      AUC
   0.3    0.7   0.8780 ←
   0.4    0.6   0.8785 ←
   0.5    0.5   0.8790 ←
   0.6    0.4   0.8794 ←
   0.7    0.3   0.8798 ←
   0.8    0.2   0.8802 ←

  Meilleur ensemble : XGB=0.8 + LGBM=0.2
  AUC Ensemble      : 0.8802
  Gain vs v6        : -0.0006
